feat: adapt game setting 'auto use item'

NateScarlet opened the pull request:
合到dev分支 master只在发新版本的时候更新

replace #326

robo-dani commented:
抱歉, 不太清楚规范, 麻烦你了

NateScarlet commented on auto_derby/single_mode/item/item.py:
```diff
@@ -455,5 +455,15 @@ def should_use_directly(self, ctx: Context) -> bool:
             return False
         return True
 
+    def auto_used(self) -> bool:
```
叫 can_be_auto_used 更好

NateScarlet commented on auto_derby/scenes/single_mode/item_menu.py:
```diff
@@ -186,6 +186,11 @@ def _select_visible_items() -> None:
+                    if match.auto_used():
+                        ctx.items.remove(match.id, remains[match.id])
+                        for _ in range(remains[match.id]):
+                            ctx.item_history.append(ctx, match)
```
我没试过开自动使用 是买的时候就用掉了还是点使用后才用掉的? 和选择界面有什么不一样

robo-dani replied:
购买完毕后 出现小人转一圈的动画, 然后进入道具选择界面时，道具自动变灰(使用了的提示), 和手动使用部分道具后的界面一样

NateScarlet replied:
我记得那个界面是选择 属性是确认后才加 如果是选择的话应该加到selected里 这样不用重复使用后逻辑

robo-dani replied:
我蠢了，我一开始是这样写， 遇到全自动使用就卡住的bug(因为没选中道具), 就没细想了。
这样我在脚本点击道具的时候加个 是否手动使用道具变量 , 加个if 套在 selected的tap上

NateScarlet commented:
这个功能应该只在 ShopExchangedItemMenu 启用 应该加个 ItemMenu._supports_auto_use 来判断

你邮件地址怎么是个 `<=>` 如果想隐藏邮箱建议用 github 提供的 noreply 邮箱  `<77256496+robo-dani@users.noreply.github.com>` (加号后面的用户名可以自定义，这个在你[设置页面](https://github.com/settings/emails)应该会显示) 不然 commit 关联不到你

robo-dani commented:
> 这个功能应该只在 ShopExchangedItemMenu 启用

有道理, 我这样是在父类里加入子类的逻辑了, 
>应该加个 ItemMenu._supports_auto_use 来判断,

直接用 `isinstance` 替代可以吗, 或者检查`scene.name`. 因为, 我希望能自动处理玩家有没有开设置, 所以想直接用类判断.
要不我在 ShopExchangedItemMenu 里重写 `item_use`, 调用完父类的之后, 通过`item_history`检查未使用的道具, 再执行一次使用道具. 

这样我先写个加if的场景判断, 假如你觉得不好, 我就在子类里重写了. 

> 你邮件地址怎么是个 <=>  

不则么用windows, git 设置错了估计.. 抱歉

robo-dani committed:
feat: adapt game setting 'auto use item'

With the new setting, items will be auto-used after exchange. So
in shop exchanged item menu scene, record item that may be disabled
by auto-used

NateScarlet commented:
为啥要用isinstance 父类不应该import子类 会循环引入

NateScarlet commented:
https://github.com/NateScarlet/auto-derby/blob/0e5027ab667bc3b1eed5cd2a6085de30a944c721/auto_derby/scenes/single_mode/item_menu.py#L108

https://github.com/NateScarlet/auto-derby/blob/0e5027ab667bc3b1eed5cd2a6085de30a944c721/auto_derby/scenes/single_mode/shop_exchanged_item_menu.py#L15

你看这

robo-dani commented:
还有个 bug, 往 select 里加item应该按照数量, 我晚上把他改掉,
